cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh

```diff
@@ -0,0 +1,297 @@
+    thrust::inclusive_scan(rmm::exec_policy(stream),
+                           num_geoms_per_pair.begin(),
+                           num_geoms_per_pair.end(),
+                           thrust::next(offsets.begin()));
+  }
+
+  /// Return range to offset array
+  auto offset_range() { return range{offsets.begin(), offsets.end()}; }
+
+  /// Return range to geometry array
+  auto geom_range() { return range{geoms.begin(), geoms.end()}; }
+
+  /// Return id_range structure to id arrays
+  auto get_id_ranges()
+  {
+    return id_ranges{range(lhs_linestring_ids.begin(), lhs_linestring_ids.end()),
+                     range(lhs_segment_ids.begin(), lhs_segment_ids.end()),
+                     range(rhs_linestring_ids.begin(), rhs_linestring_ids.end()),
+                     range(rhs_segment_ids.begin(), rhs_segment_ids.end())};
+  }
+
+  /// Return the number of pairs in the intermediates
+  auto size() { return offsets.size() - 1; }
+};
+
+/**
+ * @brief Kernel to compute intersection and store the result and look-back indices to outputs.
+ *
+ * Each thread operates on one segment in `multilinestrings1`, iterates over all segments in the
+ * other multilinestring.
+ */
+template <typename MultiLinestringRange1,
+          typename MultiLinestringRange2,
+          typename TempIt1,
+          typename TempIt2,
+          typename Offsets1,
+          typename Offsets2,
+          typename Offsets3,
+          typename IdRanges,
+          typename OutputIt1,
+          typename OutputIt2>
+void __global__ pairwise_linestring_intersection_simple(MultiLinestringRange1 multilinestrings1,
+                                                        MultiLinestringRange2 multilinestrings2,
+                                                        TempIt1 n_points_stored,
+                                                        TempIt2 n_segments_stored,
+                                                        Offsets1 num_points_offsets_first,
+                                                        Offsets2 num_segments_offsets_first,
+                                                        Offsets3 num_points_per_pair_first,
+                                                        IdRanges point_ids_range,
+                                                        IdRanges segment_ids_range,
+                                                        OutputIt1 points_first,
+                                                        OutputIt2 segments_first)
+{
+  using T       = typename MultiLinestringRange1::element_t;
+  using types_t = uint8_t;
+  using count_t = iterator_value_type<Offsets1>;
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multilinestrings1.num_points();
+       idx += gridDim.x * blockDim.x) {
+    auto const part_idx = multilinestrings1.part_idx_from_point_idx(idx);
+    if (!multilinestrings1.is_valid_segment_id(idx, part_idx)) continue;
```

Comment:
Another example of these two functions called together.

Reply:
No. I don't think they are to be combined. `part_idx` is the index to the linestring in the multilinestring. Here `idx` happens to be interpreted as the index of the segment, and for each linestring there is always an invalid segment idx at the end of the linestring. `is_valid_segment_id` detects those segment index and removes them. 

In theory the above are two logically separate steps. In reality, the second API depends on the first one. So if you want to use the second API, you always need to compute the first one. But you can have a kernel that only calls the first API and not the second.

Reply:
But do you have a kernel that only calls the first?

Reply:
> But do you have a kernel that only calls the first?

Currently no. In fact, giving this a second thought. The only case where we need to call `part_idx_from_point_idx` is when the threads are launched at the bottom of the geometry hierarchy and need to traverse the hierarchy bottom-up. When this happens we almost always need to call both function. So perhaps yes, we can merge these two helpers.

Reply:
Actually, I would add to the above: although the thread working in the geometry hierarchy bottom up, it only needs to call `is_valid_segment_id` when the thread is treating the linestring as per segment-basis. 

There are, however, other situation where we might want to treat the treat the linestring as a mere collection of points. (The recent `contains_properly` is an example for that!) In these situation we don't need to call `is_valid_segment_id` after the searching the `part_id`. It's just up to this point, APIs that uses the new `multilienstring_range` objects hasn't come across these API requirements yet.

On that note, I would recommend adding a new `part_idx_from_segment_idx`, implying that the argument is indeed id for the segment, not points. It can then return a `nullopt` if the segment id is invalid in the multilinestring.